### ISSUE TITLE
policy: incorrect policy computed for endpoint when L4-only policy and L7-policy are specified for the same port

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#3509]
Consider the following rule

```
specs:
- endpointSelector:
    matchLabels:
      id: app1
  ingress:
  - fromEndpoints:
    - matchLabels:
        id: app2
    toPorts:
    - ports:
      - port: "9093"
        protocol: TCP
      rules:
        http:
        - method: ^GET$
          path: ^/metrics$
  - toPorts:
    - ports:
      - port: "9093"
        protocol: TCP
```

In tabular form, this equates to:
```
L3      |     L4   |  L7
--------|----------|--------------------
*.      | TCP/9093 |   *
--------|----------|--------------------
app2    | TCP/9093 |   GET ^/metrics
```

Based on the Cilium policy model, we expect that to collapse to the following:

```
L3     |     L4   |  L7
-------|----------|-------------------
*.     | TCP/9093 |   *
```

In other words, the L4-only rule here shadows the L3-L4-L7 rule. However, currently, in Cilium, this forces all inbound TCP traffic on port 9093 to get forwarded to the proxy, which only allows traffic to `GET ^/metrics`. 

Resolved policy for app1 endpoint:
```
   "policy": {
      "realized": {
        "allowed-egress-identities": [],
        "allowed-ingress-identities": [
          1
        ],
        "build": 9,
        "cidr-policy": {
          "egress": [],
          "ingress": []
        },
        "id": 7846,
        "l4": {
          "egress": [],
          "ingress": [
            {
              "derived-from-rules": [
                [
                  "unspec:io.cilium.k8s.policy.name=rule1",
                  "unspec:io.cilium.k8s.policy.namespace=default"
                ],
                [
                  "unspec:io.cilium.k8s.policy.name=rule1",
                  "unspec:io.cilium.k8s.policy.namespace=default"
                ]
              ],
              "rule": "{\n  \"port\": 9093,\n  \"protocol\": \"TCP\",\n  \"l7-rules\": [\n    {\n      \"any.id=app2,k8s.io.kubernetes.pod.namespace=default\": {\n        \"http\": [\n          {\n            \"path\": \"^/metrics$\",\n            \"method\": \"^GET$\"\n          }\n        ]\n      }\n    }\n  ]\n}"
            }
          ]
        },
        "policy-revision": 9
      }
    },
...
   "proxy-statistics": [
      {
        "location": "ingress",
        "port": 80,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      },
      {
        "allocated-proxy-port": 15084,
        "location": "ingress",
        "port": 9093,
        "protocol": "http",
        "statistics": {
          "requests": {},
          "responses": {}
        }
      }
    ],
```

As one can see, all traffic is being redirected to the proxy on port 9093. Instead, no traffic should be directed to the proxy because we allow all L7 traffic implicitly with `toPorts` containing no L7-related policy, *despite the fact that there is another rule for the same port which has L7 policy*.

Change the policy resolving logic to correctly have the policy computed when L4-only policy and L3-L4-L7 policy are applied on the same port.

**Comment by rlenglet** [comment]
This is likely a bug in the code that generates the NPDS network policies.
I'll investigate.

**Comment by joestringer** [comment]
@rlenglet ah, do you think so? I was assuming it would be to do with `mergeL4Port()`.

**Comment by ianvernon** [comment]
@joestringer it is a bug in the merge logic on the Cilium side, but there is work to be done on the NPDS side as well.

**Comment by rlenglet** [comment]
What @ianvernon said. I was ignoring L4 only rules in the NPDS conversion code. I’ll send a fix shortly. This was not yet a cause of bugs (esp. not this one), but would have been an issue immediately with sidecar proxies where L7 redirects are unconditional.

**Comment by ianvernon** [comment]
I'm working on the `pkg/policy` side of work for this - assigning to myself as well.

**Comment by ianvernon** [comment]
Removing the needs-e2e test label. This is unit-testable within the Cilium-agent; we should only add e2e tests for things like CLI at runtime, agent configuration options, and whether policy is correctly plumbed into the datapath.

**Comment by tgraf** [comment]
Resolved by #3518
